Partition Blob URL fetches by Storage Key

recvfrom opened the pull request:
Partition Blob URL fetches by Storage Key

This change updates the spec to partition Blob URL fetches by Storage Key. This change is part of broader changes discussed in https://github.com/w3c/FileAPI/issues/153#issuecomment-2330085478. Specifically, we will also:
 - Partition Blob URL revocation by storage key - https://github.com/w3c/FileAPI/pull/201
 - Enforce noopener for cross-top-level-site Blob URLs - (Spec changes for this will follow)

I considered incorporating the Storage Key checks into the "resolve a blob URL" algorithm instead, but it seemed that this would require an environment settings object to be available as part of https://url.spec.whatwg.org/#url-parsing, and I'm not sure whether that is the case / a change we want.

- [x] At least two implementers are interested (and none opposed):
   * Firefox has already implemented this - https://github.com/w3c/FileAPI/issues/153#issuecomment-2332288047
   * Safari has implemented partitioning Blob URL fetches by top-level origin and is considering partitioning them by a site-based Storage Key (Storage Key specifics TBD) - https://github.com/w3c/FileAPI/issues/153#issuecomment-2332086739
- [x] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * https://chromium-review.googlesource.com/c/chromium/src/+/5967596
- [x] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chromium: https://crbug.com/40057646
   * Gecko: N/A
   * WebKit: N/A
   * Deno (not for CORS changes): N/A (it seems Deno does not partition storage or more generally implement the same-origin policy: https://docs.deno.com/runtime/reference/web_platform_apis/#spec-deviations)
- [x] [MDN issue](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) is filed: https://github.com/mdn/content/issues/36542
- [x] The top of this comment includes a [clear commit message](https://github.com/whatwg/meta/blob/main/COMMITTING.md) to use. <!-- If you created this PR from a single commit, Github copied its message. Otherwise, you need to add a commit message yourself. -->

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/fetch/1783.html" title="Last updated on Dec 6, 2024, 11:17 PM UTC (ddae2dc)">Preview</a> | <a href="https://whatpr.org/fetch/1783/1dc1b03...ddae2dc.html" title="Last updated on Dec 6, 2024, 11:17 PM UTC (ddae2dc)">Diff</a>

recvfrom commented:
@annevk would you be able to look at this PR as well? Or could you recommend another reviewer? Thank you!

domenic commented on fetch.bs:
```diff
@@ -3133,6 +3133,32 @@ or an <a>implementation-defined</a> value.
+prevented from succeeding if the <a spec=storage>storage key</a> of the
+<a>environment settings object</a> making the request is different than the
+<a spec=storage>storage key</a> of the <a>environment settings object</a> corresponding to where
+the Blob URL was created.
```
This feels like it would work better after line 5028, instead of near this algorithm.

recvfrom replied:
I was thinking this would make sense in the Infrastructure section like how the `HTTP cache partitioning` section and `determine the HTTP cache partition` algorithm are, but I've moved the algorithm under the `Scheme fetch` section now (and removed the note since it seems kinda out of place now. WDYT?

domenic commented on fetch.bs:
```diff
@@ -4978,6 +5004,29 @@ steps:
+      <p>If all of the following conditions are true:
+
+      <ul class=brief>
+       <li><p><var>request</var>'s <a for=request>mode</a> is not "<code>navigate</code>"
```
```suggestion
       <li><p><var>request</var>'s <a for=request>mode</a> is not "<code>navigate</code>";
```

domenic commented on fetch.bs:
```diff
@@ -4978,6 +5004,29 @@ steps:
+      <ul class=brief>
+       <li><p><var>request</var>'s <a for=request>mode</a> is not "<code>navigate</code>"
+
+       <li><p><var>requestStorageKey</var> is non-null
```
```suggestion
       <li><p><var>requestStorageKey</var> is non-null; and
```

domenic commented on fetch.bs:
```diff
@@ -4978,6 +5004,29 @@ steps:
+       <li><p><var>requestStorageKey</var> is non-null
+
+       <li><p><var>requestStorageKey</var> does not <a for="storage key">equal</a>
+       <var>blobStorageKey</var>
```
```suggestion
       <var>blobStorageKey</var>,
```

domenic commented on fetch.bs:
```diff
@@ -4978,6 +5004,29 @@ steps:
+     <li>
+      <p>Let <var>blobStorageKey</var> be the result of running
+      <a>obtain a storage key for non-storage purposes</a> with <var>blobURLEntry</var>'s
+      <a>environment settings object</a>.
```
This should not link to the type definition of "environment settings object", but instead link to https://w3c.github.io/FileAPI/#blob-url-entry-environment .

recvfrom replied:
ah yes thanks for pointing this out. Changed

domenic reviewed (approved):
Non-editor LGTM

domenic commented:
It looks like there aren't _too_ many references to the blob URL entry's object... https://dontcallmedom.github.io/webdex/o.html#object%40%40blob%20URL%20entry%40dfn

annevk commented:
Nice, the HTML find is legit and would also need to be updated I suspect. It would benefit from having the obtaining defined in File API so there's less need for duplication. Reporting API seems bogus: https://github.com/w3c/reporting/issues/273. Web Authentication API seems similarly bogus: https://github.com/w3c/webauthn/issues/2212.

recvfrom commented on fetch.bs:
```diff
@@ -4978,6 +4978,29 @@ steps:
       <p class=note>The `<code>GET</code>` <a for=/>method</a> restriction serves no useful purpose
       other than being interoperable.
 
+     <li>
```
Thanks, this should now be addressed

recvfrom commented:
Thanks for the feedback, I've updated my FileAPI PR [1] to expose an obtain algorithm as recommended here, and I've updated this PR to use it. One question I thought more about was what to do when the request doesn't have an associated environment... It seems like for Blob URLs we should allow navigations to bypass the partitioning checks, but otherwise we should just fail (this seems in line with the Chromium implementation at least). WDYT?

[1] https://github.com/w3c/FileAPI/pull/201/files

annevk commented on fetch.bs:
```diff
@@ -4970,15 +4970,26 @@ steps:
+     return a <a>network error</a>.
+
+     <li><p>Let <var>blob</var> be the result of <a>obtaining a blob object</a> given
+     <var>blobURLEntry</var>, <var>requestEnvironment</var>, and <var>isNavigation</var>.
```
So when isNavigation is true we don't even care about environment, right? I would merge those parameters into one.

So you have

> given an environment or the string "`navigation`" _environment_

on the other side or some such forcing callers to confront that question. And it should probably be top-level navigation?

recvfrom replied:
Thanks for the feedback - done!

annevk commented on fetch.bs:
```diff
@@ -4970,15 +4970,26 @@ steps:
+     <li><p>Let <var>blob</var> be the result of <a>obtaining a blob object</a> given
+     <var>blobURLEntry</var>, <var>requestEnvironment</var>, and <var>isNavigation</var>.
+
+     <li><p>If <var>blob</var> is failure, or <var>blob</var> is not a {{Blob}} object, then
```
You only need the second conditional to catch both conditions.

annevk commented on fetch.bs:
```diff
@@ -4970,15 +4970,26 @@ steps:
+     <a for=request>determining the environment</a> given <var>request</var>.
+
+     <li><p>Let <var>isNavigation</var> be true if <var>request</var>'s <a for=request>mode</a> is
+     not "<code>navigate</code>"; otherwise, false.
```
Perhaps instead of mode we should be checking destination for "document" which is only true for top-level navigations.

recvfrom replied:
Thanks for calling this out, I had overlooked that we didn't want to exempt all navigations from the partitioning checks and instead only want to exempt top-level navigations. I confirmed that Safari and Firefox both enforce partitioning checks on subframe navigations, so we'll implement it this way in Chromium as well, write a new WPT for this, and I'll update the spec here to look for destination equaling documenting as you suggested.

recvfrom replied:
Done!

annevk reviewed (commented):
Thank you for reworking this so quickly!

annevk commented on fetch.bs:
```diff
@@ -4970,15 +4970,32 @@ steps:
+     <li><p>Let <var>blob</var> be null.
+
+     <li><p>If <var>isTopLevelNavigation</var> is true, then set <var>blob</var> to the result of
+     <a href="https://w3c.github.io/FileAPI/#blob-url-obtain-object">obtaining a blob object</a>
```
Given that the File API PR landed we should be able to do this without `href`. Same below.

recvfrom replied:
Weird, at the time I couldn't get this to work without using href even after the File API PR spec had landed, but it does seem to work just fine now 🤔 Thanks for calling this out - changed

annevk commented on fetch.bs:
```diff
@@ -4970,15 +4970,32 @@ steps:
+
+     <li><p>If <var>isTopLevelNavigation</var> is true, then set <var>blob</var> to the result of
+     <a href="https://w3c.github.io/FileAPI/#blob-url-obtain-object">obtaining a blob object</a>
+     given <var>blobURLEntry</var> and the string "<code>navigation</code>".
```
I think it would be clearer to first initialize a variable called _navigationOrEnvironment_ and then invoke "obtaining a blob object" once. At least I find code written that way easier to follow.

recvfrom replied:
Thanks for the feedback, I wasn't sure if it was OK to assign values of different types into a variable. It's indeed simpler with only the one all to "obtaining a blob object". Thanks, and done!

annevk reviewed (commented):
Thanks again for improving this aspect of `blob:` URLs!

recvfrom commented:
> Thanks again for improving this aspect of `blob:` URLs!

NP, thanks for all of your helpful review feedback and guidance!
